test/Streamly/Test/Data/Unbox.hs

```diff
@@ -0,0 +1,86 @@
+{-# OPTIONS_GHC -Wno-missing-methods #-}
+-- |
+-- Module      : Streamly.Test.Data.Unbox
+-- Copyright   : (c) 2019 Composewell technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+module Streamly.Test.Data.Unbox (main) where
+
+import Data.Proxy
+import Foreign.Storable (Storable(..))
+import GHC.Generics hiding (moduleName)
+import Streamly.Internal.Data.Unboxed
+    ( SizeOfRep
+    , Unbox(..)
+    , genericPeekByteIndex
+    , genericPokeByteIndex
+    , genericSizeOf
+    )
+import qualified Streamly.Internal.Data.IORef.Unboxed as IORef
+
+import Test.Hspec as H
+import Test.Hspec.QuickCheck
+import Test.QuickCheck (Property)
+import Test.QuickCheck.Monadic (monadicIO, assert, run)
+
+#define DERIVE_UNBOX(type) \
+instance Storable type where \
+    { sizeOf _ = genericSizeOf (Proxy :: Proxy type) }; \
+\
+instance Unbox type where { \
+    peekByteIndex = genericPeekByteIndex; \
+    pokeByteIndex = genericPokeByteIndex }
+
+test ::(Unbox a, Eq a) => a -> a -> Property
+test v1 v2 = monadicIO $ do
+    ref <- run $ IORef.newIORef v1
+    x1 <- run $ IORef.readIORef ref
+    run $  IORef.writeIORef ref v2
+    x2 <- run $  IORef.readIORef ref
+    assert (x2 == v2)
+    assert (x1 == v1)
+
+checkSizeOf :: (SizeOfRep (Rep a)) => Proxy a -> Int -> Property
+checkSizeOf v size = monadicIO $
+    assert (genericSizeOf v == size)
+
+data Unit = Unit deriving (Show, Generic, Eq)
+DERIVE_UNBOX(Unit)
```

Comment:
Now that we have default methods. We can use `instance Unbox Unit` here and remove the DERIVE_UNBOX macro.